apis/v1alpha1/tlsroute_types.go

```diff
@@ -0,0 +1,131 @@
+}
+
+// TLSRouteRule is the configuration for a given rule.
+type TLSRouteRule struct {
+	// Match defines which connections match this rule.
+	Match *TLSRouteMatch `json:"match" protobuf:"bytes,1,opt,name=match"`
+	// Action defines what happens to the connection.
+	Action *TLSRouteAction `json:"action" protobuf:"bytes,2,opt,name=action"`
+}
+
+// TLSRouteAction is the action for a given rule.
+type TLSRouteAction struct {
+	// ForwardTo sends requests to the referenced object(s). The
+	// resource may be "services" (omit or use the empty string for the
+	// group), or an implementation may support other resources (for
+	// example, resource "myroutetargets" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "services".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the HTTPRoute will be true.
+	//
+	// Support: core
+	//
+	// +kubebuilder:validation:MinItems=1
+	ForwardTo []ForwardToTarget `json:"forwardTo" protobuf:"bytes,1,rep,name=forwardTo"`
+
+	// ExtensionRef is an optional, implementation-specific extension to the
+	// "action" behavior.  The resource may be "configmaps" (use the empty
+	// string for the group) or an implementation-defined resource (for
+	// example, resource "myrouteactions" in group "networking.acme.io").
+	// Omitting or specifying the empty string for both the resource and
+	// group indicates that the resource is "configmaps".  If the referent
+	// cannot be found, the "InvalidRoutes" status condition on any Gateway
+	// that includes the TLSRoute will be true.
+	//
+	// Support: custom
+	//
+	// +optional
+	ExtensionRef *RouteActionExtensionObjectReference `json:"extensionRef,omitempty" protobuf:"bytes,2,opt,name=extensionRef"`
+}
+
+// TLSRouteMatch defines the predicate used to match connections to a
+// given action.
+type TLSRouteMatch struct {
+	// SNIs defines a set of SNI names that should match against the
+	// SNI attribute of TLS CLientHello message in TLS handshake.
+	//
+	// SNI can be "precise" which is a domain name without the terminating
+	// dot of a network host (e.g. "foo.example.com") or "wildcard", which is
+	// a domain name prefixed with a single wildcard label (e.g. "*.example.com").
+	// The wildcard character '*' must appear by itself as the first DNS
+	// label and matches only a single label.
+	// You cannot have a wildcard label by itself (e.g. Host == "*").
+	// Requests will be matched against the Host field in the following order:
+	// 1. If SNI is precise, the request matches this rule if
+	//    the SNI in ClientHello is equal to one of the defined SNIs.
+	// 2. If SNI is a wildcard, then the request matches this rule if
+	//    the SNI is to equal to the suffix
+	//    (removing the first label) of the wildcard rule.
+	//
+	// Support: core
+	//
+	SNIs []string `json:"snis,omitempty" protobuf:"bytes,1,rep,name=snis"`
```

Comment:
We had feedback on the API doc that there are multiple semantically-equivalent ways to match hostnames . Can we use `HostnameMatch` here?

Reply:
I did consider that feedback and I think Dan has a great point.
I decided to follow HTTPRoute in this case as this is also a route.

I initially started with changing HTTPRoute to include HostnameMatch in a different PR but that feels more verbose than it needs to be. We should instead change listener in G/W to mirror what we have in HTTPRoute.

Reply:
Ok let's keep this for now and table the wider discussion.